Add CBMC proof for default_cmm_generate_enc_materials

adpaco-aws opened the pull request:
This PR adds a CBMC proof for default_cmm_generate_enc_materials, a function from default_cmm.c,

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

# Check any applicable:
- [ ] Were any files moved? Moving files changes their URL, which breaks all hyperlinks to the files.

alex-chew commented on verification/cbmc/proofs/default_cmm_generate_enc_materials/default_cmm_generate_enc_materials_harness.c:
```diff
@@ -0,0 +1,116 @@
+
+    enum aws_cryptosdk_alg_id alg_id;
+    if (nondet_bool()) request->requested_alg = alg_id;
+    // We can either add this OR add "if (!props) goto err;" after line 61 in the source code
+    // I think adding the check to the source code is better in this case.
+    // struct aws_cryptosdk_alg_properties *alg_props = aws_cryptosdk_alg_props(alg_id);
+    // __CPROVER_assume(aws_cryptosdk_alg_properties_is_valid(alg_props));
```
I agree that the check in the function implementation is better.

```suggestion
    // The function ensures that `aws_cryptosdk_alg_props(alg_id)` is valid.
```

feliperodri replied:
I'm not really sure I understand this assumption. Is this related tot he next assumption that `cmm` has a set `alg_id`? Why do you even need the next assumption anyway?

tegansb replied:
I also like the `if (!props) goto err;` check in the function for this.

alex-chew reviewed (changes requested):
LGTM, pending the comment about the `aws_cryptosdk_alg_props(alg_id)` validity assertion.

feliperodri commented on verification/cbmc/proofs/default_cmm_generate_enc_materials/default_cmm_generate_enc_materials_harness.c:
```diff
@@ -0,0 +1,116 @@
+
+    /* Operation under verification */
+    if (__CPROVER_file_local_default_cmm_c_default_cmm_generate_enc_materials(cmm, output, request) == AWS_OP_SUCCESS) {
+        assert(aws_cryptosdk_enc_materials_is_valid(*output));
```
Is there any extra properties you can assert over `output`, `cmm`, or `request` in case of success or failure? @adpaco-aws @alex-chew

alex-chew replied:
If successful, you can verify:

* `request->requested_alg` is a valid algorithm ID

If the call fails, you can verify:

* `request->requested_alg` is a valid algorithm ID or zero
* `output` is unchanged

feliperodri commented:
/cbmc run checks

alex-chew commented on verification/cbmc/proofs/default_cmm_generate_enc_materials/default_cmm_generate_enc_materials_harness.c:
```diff
@@ -0,0 +1,116 @@
+    // __CPROVER_assume(aws_cryptosdk_alg_properties_is_valid(alg_props));
+
+    /* Nondet set self->default_alg to a valid alg_id */
+    __CPROVER_assume(aws_cryptosdk_default_cmm_set_alg_id(cmm, alg_id) == AWS_OP_SUCCESS);
```
This misses the edge case where the CMM's `default_alg` is unset, i.e. it has the special value `DEFAULT_ALG_UNSET` (defined in `default_cmm.c`) which is not considered a valid algorithm: https://github.com/aws/aws-encryption-sdk-c/blob/d6ea33e9390960737d3c3888b014bbe229b42953/source/default_cmm.c#L176-L178

adpaco-aws replied:
Please see the comment below:

> Sorry, there were two things wrong in here:
> * `aws_cryptsodk_algorithm_is_known` was not defined (fixed by adding header.c to Makefile).
> * Using an the `cmm` initializer instead of a nondet. allocator (which I have included now).
> So this code has been removed.

Because of this, all cases are now covered.

adpaco-aws commented on verification/cbmc/proofs/default_cmm_generate_enc_materials/default_cmm_generate_enc_materials_harness.c:
```diff
@@ -0,0 +1,116 @@
+void default_cmm_generate_enc_materials_harness() {
+    /* Nondet input required to init cmm */
+    struct aws_cryptosdk_keyring *keyring        = malloc(sizeof(*keyring));
+    const struct aws_cryptosdk_keyring_vt vtable = { .vt_size    = sizeof(struct aws_cryptosdk_keyring_vt),
```
Yes, we have it now.

adpaco-aws commented on verification/cbmc/proofs/default_cmm_generate_enc_materials/default_cmm_generate_enc_materials_harness.c:
```diff
@@ -0,0 +1,116 @@
+    __CPROVER_assume(keyring->vtable != NULL);
+
+    /* Nondet input */
+    struct aws_cryptosdk_cmm *cmm               = aws_cryptosdk_default_cmm_new(can_fail_allocator(), keyring);
```
I have created a new allocator function for default_cmm (which is an extension of cmm). We are not using this function anymore.

adpaco-aws commented on verification/cbmc/proofs/default_cmm_generate_enc_materials/default_cmm_generate_enc_materials_harness.c:
```diff
@@ -0,0 +1,116 @@
+
+    /* Nondet input */
+    struct aws_cryptosdk_cmm *cmm               = aws_cryptosdk_default_cmm_new(can_fail_allocator(), keyring);
+    struct aws_cryptosdk_enc_materials **output = malloc(sizeof(*output));
+    struct aws_cryptosdk_enc_request *request   = malloc(sizeof(*request));
+
+    /* Assumptions */
+    __CPROVER_assume(cmm != NULL);
+
+    __CPROVER_assume(output != NULL);
+
+    __CPROVER_assume(request != NULL);
+    request->alloc = can_fail_allocator();
+    __CPROVER_assume(aws_allocator_is_valid(request->alloc));
+    request->enc_ctx = malloc(sizeof(*request->enc_ctx));
+    __CPROVER_assume(request->enc_ctx != NULL);
+    ensure_allocated_hash_table(request->enc_ctx, MAX_NUM_ITEMS);
+    __CPROVER_assume(aws_cryptosdk_enc_request_is_valid(request));
```
Added them.

adpaco-aws reviewed (commented):
Sorry, I had this PR almost updated but I wanted to inspect some stuff in detail. We have a new solution which includes as much feedback from you when possible.

adpaco-aws commented:
@alex-chew I have tried the first one in here
> If the call fails, you can verify:
>
>    request->requested_alg is a valid algorithm ID or zero
>    output is unchanged
```
assert(!request->requested_alg || aws_cryptosdk_algorithm_is_known(request->requested_alg);
```
and
```
assert(request->requested_alg == DEFAULT_ALG_UNSET ||
           aws_cryptosdk_algorithm_is_known(request->requested_alg));
```
(in case you meant `DEFAULT_ALG_UNSET` as zero). But none of them work for me. Can you please elaborate?

Also, I am not sure how to check that `output` remains unchanged. Do we have templates for this? @feliperodri @tegansb

alex-chew commented on source/default_cmm.c:
```diff
@@ -37,6 +37,10 @@ static int default_cmm_generate_enc_materials(
     struct aws_cryptosdk_enc_materials **output,
     struct aws_cryptosdk_enc_request *request) {
+    AWS_PRECONDITION(cmm != NULL);
+    AWS_PRECONDITION(output != NULL);
```
You are correct that `output` is an out parameter for the generated encryption materials, and at the start of the function it is typically not the case that `**output` is a valid encryption materials struct. A stronger precondition that would make sense might be to assert that `output` is writable for `sizeof(*output)` bytes.

alex-chew commented:
> @alex-chew I have tried the first one in here
> 
> > If the call fails, you can verify:
> > request->requested_alg is a valid algorithm ID or zero
> > output is unchanged
> 
> ```
> assert(!request->requested_alg || aws_cryptosdk_algorithm_is_known(request->requested_alg);
> ```
> 
> and
> 
> ```
> assert(request->requested_alg == DEFAULT_ALG_UNSET ||
>            aws_cryptosdk_algorithm_is_known(request->requested_alg));
> ```
> 
> (in case you meant `DEFAULT_ALG_UNSET` as zero). But none of them work for me. Can you please elaborate?

Oops, that's my bad. If the call fails, it should be the case that `request->requested_alg` is valid or *unchanged* (not zero). Sorry for the confusion. In retrospect, however, I'm not sure this is worth verifying.

> Also, I am not sure how to check that `output` remains unchanged. Do we have templates for this? @feliperodri @tegansb

You can save its value prior to calling the function under verification and check that its value is unchanged by the call. e.g. https://github.com/aws/aws-encryption-sdk-c/blob/master/verification/cbmc/proofs/aws_cryptosdk_priv_hdr_parse_alg_id/aws_cryptosdk_priv_hdr_parse_alg_id_harness.c

adpaco-aws commented:
@alex-chew Ok, I have removed the assert in case of failure.

I have also added code to check that output remains unchanged at the top level (i.e., it performs a cast of the struct and compares the values as an array, but does not check for changes on structures being pointed to by its members). The main reason to leave it as it is performance, since doing these checks would be time-consuming and the structures are not really used in the function under verification (remember that `output` is just a placeholder in this function).

feliperodri commented on source/default_cmm.c:
```diff
@@ -180,3 +185,10 @@ int aws_cryptosdk_default_cmm_set_alg_id(struct aws_cryptosdk_cmm *cmm, enum aws
 }
+
+bool aws_cryptosdk_default_cmm_is_valid(const struct aws_cryptosdk_cmm *cmm) {
+    AWS_PRECONDITION(cmm != NULL);
```
It should return false in case `cmm` is NULL, not fail on an assertion.

adpaco-aws replied:
Right, I am happy you saw this because I had my doubts about it.

feliperodri commented on verification/cbmc/proofs/default_cmm_generate_enc_materials/default_cmm_generate_enc_materials_harness.c:
```diff
@@ -0,0 +1,71 @@
+    struct aws_cryptosdk_enc_request *request  = ensure_enc_request_attempt_allocation(MAX_TABLE_SIZE);
+
+    /* Assumptions */
+    __CPROVER_assume(cmm != NULL);
+    __CPROVER_assume(aws_cryptosdk_default_cmm_is_valid(cmm));
```
You wouldn't need it if `aws_cryptosdk_default_cmm_is_valid` returned false in case `cmm` is NULL.

adpaco-aws replied:
Removed null assumption, thanks!

feliperodri commented on verification/cbmc/proofs/default_cmm_generate_enc_materials/default_cmm_generate_enc_materials_harness.c:
```diff
@@ -0,0 +1,71 @@
+
+    __CPROVER_assume(output != NULL);
+
+    __CPROVER_assume(request != NULL);
+    __CPROVER_assume(aws_allocator_is_valid(request->alloc));
+    __CPROVER_assume(request->enc_ctx != NULL);
+    __CPROVER_assume(aws_cryptosdk_enc_request_is_valid(request));
```
This looks odd to me too. Why the first three assumptions are not part of the validity check for `aws_cryptosdk_enc_request`?

adpaco-aws replied:
This is probably from my early days of proof-writing. The assumptions were not needed so I removed them.

feliperodri commented on verification/cbmc/sources/make_common_data_structures.c:
```diff
@@ -30,6 +31,14 @@
+    struct aws_cryptosdk_cmm base;
+    struct aws_allocator *alloc;
+    struct aws_cryptosdk_keyring *kr;
+    // Invariant: this is either DEFAULT_ALG_UNSET or is a valid algorithm ID
```
Avoid cpp comments.
```c
/* Invariant: this is either DEFAULT_ALG_UNSET or is a valid algorithm ID. */
```

adpaco-aws replied:
I copied this from the source code, so I fixed it in both places.

feliperodri commented on verification/cbmc/sources/make_common_data_structures.c:
```diff
@@ -253,3 +262,50 @@ void ensure_nondet_allocate_cmm_vtable_members(struct aws_cryptosdk_cmm_vt *vtab
+    /* Nondet input required to init cmm */
+    struct aws_cryptosdk_keyring *keyring = malloc(sizeof(*keyring));
+
+    /* Assumptions required to init cmm */
+    ensure_cryptosdk_keyring_has_allocated_members(keyring, vtable);
+    __CPROVER_assume(aws_cryptosdk_keyring_is_valid(keyring));
+    __CPROVER_assume(keyring->vtable != NULL);
```
Create a GitHub issue to investigate a way to move these assumptions to the validity function and not to the allocator.

adpaco-aws replied:
Opened one here: #675

feliperodri commented on verification/cbmc/sources/make_common_data_structures.c:
```diff
@@ -253,3 +262,50 @@ void ensure_nondet_allocate_cmm_vtable_members(struct aws_cryptosdk_cmm_vt *vtab
+    struct default_cmm *self = NULL;
+    if (cmm) {
+        self        = (struct default_cmm *)cmm;
+        self->alloc = can_fail_allocator();
```
Could it be NULL? We should be as nondet as possible here.

adpaco-aws replied:
Yes, I missed that one. It can be `NULL` now.

feliperodri commented on verification/cbmc/stubs/on_encrypt_stub.c:
```diff
@@ -0,0 +1,74 @@
+         */
+        if (nondet_bool()) ensure_byte_buf_has_allocated_buffer_member(unencrypted_data_key);
+    }
+    int ret;
+    return ret;
```
```c
return nondet_int();
```

adpaco-aws replied:
Done as well.

alex-chew reviewed (approved):
LGTM, pending CI fixes.

feliperodri commented on source/default_cmm.c:
```diff
@@ -180,3 +185,11 @@ int aws_cryptosdk_default_cmm_set_alg_id(struct aws_cryptosdk_cmm *cmm, enum aws
     return AWS_OP_SUCCESS;
 }
+
+bool aws_cryptosdk_default_cmm_is_valid(const struct aws_cryptosdk_cmm *cmm) {
+    if (cmm == NULL) return false;
+
+    struct default_cmm *self = (struct default_cmm *)cmm;
+    return aws_cryptosdk_cmm_base_is_valid(&self->base) && aws_allocator_is_valid(self->alloc) &&
+           aws_cryptosdk_keyring_is_valid(self->kr);
+}
```
Why this function wasn't written like...
```c
bool default_cmm_is_valid(const struct default_cmm *cmm) {
    return (cmm != NULL) &&
           aws_cryptosdk_cmm_base_is_valid(&cmm->base) &&
           aws_allocator_is_valid(cmm->alloc) &&
           aws_cryptosdk_keyring_is_valid(cmm->kr);
}
```
Shouldn't we validate both structures separately (i.e., on validity function for each data structure)?

adpaco-aws replied:
In general, extensions of `aws_cryptosdk_cmm` are always handled like this in the source code: The type is used as a parameter/return type and the struct is cast before doing any operations on it. This way, we are following API guidelines.

I am not sure what you mean by "validate both structures separately". We have `aws_cryptosdk_cmm_base_is_valid` for validating `aws_cryptosdk_cmm` structs and this one to validate `default_cmm` structs.

feliperodri commented on verification/cbmc/sources/make_common_data_structures.c:
```diff
@@ -369,4 +379,51 @@ struct aws_cryptosdk_dec_materials *dec_materials_setup(
+        cmm->vtable = vtable;
+    }
+
+    struct default_cmm *self = NULL;
+    if (cmm) {
+        self        = (struct default_cmm *)cmm;
+        self->alloc = nondet_bool() ? NULL : can_fail_allocator();
+        self->kr    = keyring;
+    }
+    return (struct aws_cryptosdk_cmm *)self;
```
Why?

feliperodri commented on verification/cbmc/stubs/on_encrypt_stub.c:
```diff
@@ -0,0 +1,73 @@
+ * limitations under the License.
+ */
+
+#include <stdlib.h>
```
Why do you need that include?

adpaco-aws replied:
You are right, this was not required.

feliperodri reviewed (approved):
I have a few more questions, but no blockers.
